### PR TITLE
Meets #18200: Hide and disallow changing of administrative menus

### PR DESCRIPTION
**Pull request by netfighter**
[`* `#18200` Hide and disallow changing of administrative menus`](https://community.openproject.org/work_packages/18200)

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://netfighter-feature-18200-hide-and-disallow-admin-menu.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://netfighter-feature-18200-hide-and-disallow-admin-menu.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by machisuji** [comment]
@netfighter can you make it possible to override the hidden menu items through env variables? We could use a space separated list for the env value. For this the code would have to be able to deal with both an actual array and said string. I suggest adding a helper method for that.

File: lib/open_project/configuration.rb [hunk 1]
```diff
@@ -74,7 +74,8 @@ module Configuration
       'disable_password_login' => false,
       'omniauth_direct_login_provider' => nil,
 
-      'disable_password_choice' => false
+      'disable_password_choice' => false,
+      'hidden_menu_items' => []
```
Comment (machisuji) [inline lib/open_project/configuration.rb]: Shouldn't the default value be a hash (`{}`) then?

Comment (netfighter) [inline lib/open_project/configuration.rb]: yep, will change it...

**Comment by machisuji** [comment]
@netfighter @linki It would actually be enough to support a way to override array values through env vars. We can already do hashes. 

For instance:

``` yaml
production:
  hidden_menu_items:
    admin_menu:
      - roles
      - types
    projects_menu:
      - timelines
      - budgets
```

Say we support space (or newline) separated lists in ENV vars we could override that through the environment like this already:

``` bash
OPENPROJECT_HIDDEN__MENU__ITEMS_ADMIN__MENU='roles types'
OPENPROJECT_HIDDEN__MENU__ITEMS_PROJECTS__MENU='timelines budgets'
```

Hash support has already been added to support nested entries in `configuration.yml`.

One could add a helper to [helpers.rb](https://github.com/opf/openproject/blob/dev/lib/open_project/configuration/helpers.rb) that wraps the hidden menu items hash then and substitutes the strings for arrays if they got spaces (or newlines) in them.

Something like:

``` ruby
module OpenProject
  module Configuration
    module Helpers
      # overrides default getter in OpenProject::Configuration
      def hidden_menu_items
        menus = self['hidden_menu_items'].map do |label, nodes|
          if nodes =~ / /
            [label, nodes.split(' ')]
          else
            [label, nodes]
          end
        end
        Hash[menus]
      end
    end
  end
end
```

Though one would probably want to do that only once (i.e. cache the substitute).

File: lib/redmine/menu_manager/menu_helper.rb [hunk 2]
```diff
@@ -265,4 +265,13 @@ def allowed_node?(node, user, project)
       return true
     end
   end
+
+  def visible_node?(menu, node)
+    if OpenProject::Configuration['hidden_menu_items'].length > 0
+      hidden_nodes = OpenProject::Configuration['hidden_menu_items'][menu.to_s] || []
```
Comment (machisuji) [inline lib/redmine/menu_manager/menu_helper.rb]: Also, you should use the getter here. I.e. `OpenProject::Configuration.hidden_menu_items`.

**Comment by machisuji** [comment]
:+1:

**Comment by slawa85** [comment]
:+1:

**Pull request merged**